File: desktop-app-v3/src/lib/blocking.ts
import { create } from 'zustand';
import { invoke } from '@tauri-apps/api/core';

interface BlockingState {
  /** True if the FlowShield-managed region is currently in /etc/hosts. */
  active: boolean;
  /** True while an apply/clear elevation prompt is in flight. */
  busy: boolean;
  error: string | null;

  /** Cheap read-only sync — `blocking_status` Tauri command. No prompt. */
  refresh: () => Promise<void>;
  /** Spawns the privileged child to write the hosts region. Prompts. */
  apply: (domains: string[]) => Promise<void>;
  /** Spawns the privileged child to remove the region. Prompts. */
  clear: () => Promise<void>;
}

function errorMessage(err: unknown, fallback: string): string {
  if (typeof err === 'string') return err;
  if (err && typeof err === 'object') {
    const obj = err as { message?: string; error?: string };
    return obj.message ?? obj.error ?? fallback;
  }
  return fallback;
}

export const useBlockingStore = create<BlockingState>((set) => ({
  active: false,
  busy: false,
  error: null,

  refresh: async () => {
    try {
      const active = await invoke<boolean>('blocking_status');
      set({ active, error: null });
    } catch (err) {
      set({ error: errorMessage(err, 'Failed to read blocking status') });
    }
  },

  apply: async (domains) => {
    set({ busy: true, error: null });
    try {
      await invoke('blocking_apply', { domains });
      set({ active: true, busy: false });
    } catch (err) {
      set({
        error: errorMessage(err, 'Failed to apply blocks'),
        busy: false,
      });
      throw err;
    }
  },

  clear: async () => {
    set({ busy: true, error: null });
    try {
      await invoke('blocking_clear');
      set({ active: false, busy: false });
    } catch (err) {
      set({
        error: errorMessage(err, 'Failed to clear blocks'),
        busy: false,
      });
      throw err;
    }
  },
}));

File: desktop-app-v3/src-tauri/src/commands/blocking.rs
//! Hosts-file blocking commands. The frontend invokes these to enter
//! deep-work mode and to release the block on session end.
//!
//! These commands re-invoke the binary itself with an elevation prompt
//! (Linux: pkexec; macOS: osascript) so the GUI never holds root. The
//! privileged child runs `flowshield_desktop_lib::run_blocker_subcommand`
//! which dispatches `--blocking-apply <domains>` / `--blocking-clear`
//! and exits.
//!
//! Wrapped in `spawn_blocking` because the elevation subprocess is
//! synchronous and we don't want to stall the async runtime while the
//! user types their password.

use super::elevation;
use crate::error::{AppError, AppResult};

#[tauri::command]
pub async fn blocking_apply(domains: Vec<String>) -> AppResult<()> {
    // The frontend hands us the user's `primaryDistractions` list verbatim,
    // which is a list of CATEGORIES (e.g. "Social Media") — not real
    // domains. Expand here to the actual domains before paying the
    // elevation prompt; the privileged child stays dumb (just writes
    // whatever we tell it). Already-real domains pass through unchanged
    // so a future custom-domains UI works without further plumbing.
    let domains = crate::blocking::expand_categories(&domains);
    if domains.is_empty() {
        // Nothing to block (categories all unmapped + no raw domains) —
        // short-circuit so we don't pop the elevation dialog for a no-op.
        return Ok(());
    }
    // Comma-join because we pass through pkexec/osascript argv, where
    // each token is a separate shell argument; one packed string keeps
    // the wire format simple. The privileged child splits on commas.
    let arg = domains.join(",");
    tauri::async_runtime::spawn_blocking(move || {
        elevation::run_self_elevated("--blocking-apply", vec![arg])
    })
    .await
    .map_err(|e| AppError::Storage(format!("blocking task: {e}")))?
}

#[tauri::command]
pub async fn blocking_clear() -> AppResult<()> {
    tauri::async_runtime::spawn_blocking(|| {
        elevation::run_self_elevated("--blocking-clear", vec![])
    })
    .await
    .map_err(|e| AppError::Storage(format!("blocking task: {e}")))?
}

/// Read-only check (no elevation): is the FlowShield block region
/// currently in the hosts file? The frontend uses this to sync the
/// dashboard's "blocking" indicator on launch — picks up state left
/// over from a previous run that ended uncleanly.
#[tauri::command]
pub async fn blocking_status() -> AppResult<bool> {
    tauri::async_runtime::spawn_blocking(crate::blocking::is_active)
        .await
        .map_err(|e| AppError::Storage(format!("blocking task: {e}")))?
}
